stdlib/dynarray.mli

```diff
@@ -0,0 +1,255 @@
+    The first element is used as the filler.
+
+    This is a more advanced operation that is only useful for performance
+    purposes.
+
+    @raise Invalid_arg if the array is empty or
+      if the size is not suitable (negative, or too big for OCaml arrays)
+*)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0].
+
+    Note that an empty dynarray might still have non-0 underlying capacity
+    and therefore non-0 memory footprint. *)
+
+val add_last : 'a t -> 'a -> unit
+(** [add_last a x] adds the element [x] at the end of the array [a].
+
+    This might grow the underlying storage of [a] if it is full.
+
+    Calling [add_last a] n times is amortized O(n) complexity,
+    and O(ln(n)) reallocations of the underlying array. *)
+
+val unsafe_add_last : 'a t -> 'a -> unit
+(** [unsafe_add_last a x] adds [x] as the last element of [a],
+    assuming there is room for it in [a] already
+    (e.g. using {!ensure_capacity_with}).
+
+    It is unspecified what happens if the capacity is not enough.
+    This is for advanced use cases only. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] adds all elements of [b] at the end of [a],
+    in the order they appear in [b]. [b] is not modified.
+
+    For example, [a] will contain [1,2,3,4,5,6] after this code runs:
+    {[
+      let a = of_list [1;2;3];;
+      let b = of_list [4;5;6];;
+      let () = append a b;;
+    ]}
+    *)
+
+val append_array : 'a t -> 'a array -> unit
+(** Like {!append}, with an array. *)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append} but with an iterator. *)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append} but with a list. *)
+
+val append_iter :
+  'a t ->
+  (('a -> unit) -> 'iter-> unit) ->
+  'iter -> unit
+(** [append_iter a iter x] adds to [a] each element in [x]. It uses [iter]
+    to iterate over [x].
+
+    For example, [append_iter a List.iter [1;2;3]] would add elements
+    [1], [2], and [3] at the end of [a].
+    [append_iter a Queue.iter q] adds elements from the queue [q]. *)
+
+val pop_last_opt : 'a t -> 'a option
+(** [pop_last_opt a] removes and returns the last element of [a],
+    or [None] if the array is empty. *)
+
+val pop_last : 'a t -> 'a
+(** [pop_last a] removes and returns the last element of [a], assuming
+    [a] is not empty.
+    @raise Not_found on an empty array. *)
+
+val remove_last : 'a t -> unit
+(** [remove_last a] removes the last element of [a], or does nothing
+    if [is_empty a]. *)
+
+val copy : 'a t -> 'a t
+(** [copy a] is a shallow copy of [a], that can be modified independently. *)
+
+val truncate : 'a t -> int -> unit
+(** [truncate a n] truncates [a] to have at most [n] elements.
+
+    It removes elements whose index is great or equal than [n].
+    It does nothing if [n >= length a].
+
+    It is similar to:
+    {[
+      while length a > n do
+        remove_last a
+      done
+    ]} *)
+
+val fit_capacity : 'a t -> unit
```

Comment:
Since there is no prior art in OCaml's standard library yet, it might be worth using the same function name as in some other language rather than inventing yet another new name users will have to remember. Java uses `trim_to_size`, while C++ and Rust use `shrink_to_fit`